### DISCUSSION TITLE
Sensors for depth and velocity

### DISCUSSION THREAD

**Discussion by Bit1000** [FakeFishGames/Barotrauma#6792] [Ideas and suggestions]
**A wider range of sensors/detectors would open the door (no pun intended) for a lot more ideas and testing, and I think ones for temporospatial purposes would be particularly valuable.**

There are plenty of very useful electrical components for designing fairly complex control circuits (boolean gates, arithmetic/trigonometric operations, comparators, etc.) in the game already, but sadly, there are not a lot of ways to feed variables into them that would be very useful for creating feedback for things (e.g. the engine).

It is clear that the variables for depth and velocity already exist and are displayed in the game, as the Navigation Terminal readout shows vertical + horizontal velocities, as well as depth. What I find disappointing is that there are no connections on the terminal to output those values. I think that, at the very least, adding in connections for outputting those 3 values to be wired out to a control circuit would be very useful.

What I think would be even better, however, would be to break those functions out into discrete components. If there was a depth sensor, a feedback loop could be designed for a ballast tank that could run independently of the Navigation Terminal (e.g. for holding a specific depth). Similarly, I have a design in mind for an "emergency brake" button that can be pushed for the engine (requiring the horizontal velocity value) or ballast tanks (requiring the vertical velocity value), rather than having to flip the steering to autopilot and setting "maintain position." I know some will say "just use the Navigation Terminal," but there are cases where it would be helpful to have that sort of thing able to run automatically when isolated -- say, in Drone Subs, or in extremely large vessels where you have to run a massive distance to get to the main control room. There are situations where someone on the opposite end of the ship might want to be able to see the current depth/velocity and/or initiate an emergency stop.

These components (depth, x_velocity, y_velocity -- or even a single combined velocity component that has x and y outputs) could be added to every sub as a required piece of gear which is wired to their Navigation Terminals, rather than so much of the automation being "baked-in" to the Navigation Terminal as its own standalone sort of black-box computer that can't give many of its vars as outputs or receive many inputs. It'd be nice to see more submarine editors making use of what seems like a rather robust logic-based control system. Presently, there are just really simple things like water detectors and smoke detectors being joined together with "or" gates to link to a light, with no real math actually being performed -- only basic signal junction. Likewise with motion detectors generally only being used to control a hatch in most vanilla sub designs. The only other detector that seems to exist beside those is the oxygen detector, which I have yet to see actually implemented in any design (and seems quite niche).

**Comment by MaaiG** [comment]
>What I think would be even better, however, would be to break those functions out into discrete components. 

I’ve been thinking about that, too. It would be helpful. At least that information is needed in the struggle for survival. When falling into the abyss it is necessary to know exactly when to continue fighting and when to run to the escape pod. ))

>or in extremely large vessels where you have to run a massive distance to get to the main control room. 

On the big boat, I duplicate the emergency control system in the rescue room.

>There are situations where someone on the opposite end of the ship might want to be able to see the current depth/velocity 

That’s why I set up emergency posts in several places

> and/or initiate an emergency stop.

Only the captain and first officer have such authority. ))

>The only other detector that seems to exist beside those is the oxygen detector

Now this is a pretty stupid detector. Even when the hatch is open and the airlock chamber is filled with water, the detector shows 99-100% oxygen. Until the game learns to remove air from an open airlock, this detector has no interesting application IMHO.
...But it can turn on and off the oxygen generator to save battery power. Energy surge are attached ))

**Reply by mLuby** [reply]
Happy to report that as of the _Among the Ancients_ update, (Shuttle) Navigation Terminals have pins for `current_velocity_x`, `current_velocity_y`, `current_position_x`, and `current_position_y`. And I've [confirmed that it makes emergency braking easy](https://steamcommunity.com/sharedfiles/filedetails/?id=2644581764), quite helpful for sea bikes.

Personally I'd prefer if those pins were on a less cluttered device. The Motion Detector is a good candidate, at least for the `current_velocity_x/y` pins: it'd be like an accelerometer or IMU.

The Water Sensor could reasonably act as a barometer, where instead of the `high_pressure` pin outputting 0 or 1 it could output 0 or the "current depth" (`current_position_y`).

 In the real world, longitude (`current_position_x`) is more difficult to determine; perhaps that explains the need for Beacon Stations all over Europa.

**Comment by MaaiG** [comment]
>The Water Sensor could reasonably act as a barometer

Exactly!

>Personally I'd prefer if those pins were on a less cluttered device.

And at the very least this way:
![143297624-d32b9325-be18-4f4f-ad93-a2a67ab0bfed (1)](https://user-images.githubusercontent.com/93348499/143501249-e6a613f7-60f4-4c1e-8ae4-5f3345e5c431.png)
